src/components/WalletUnlock/DeterministicAccountList.tsx

```diff
@@ -88,87 +80,120 @@ const SDownloader = styled(Downloader)`
+      $total: () => maxEmptyAddresses
+    }}
+  />
+);
 
 export default function DeterministicAccountList({
   finishedAccounts,
   asset,
   isComplete,
-  onUnlock,
   network,
-  freshAddressIndex,
-  generateFreshAddress,
+  displayEmptyAddresses,
+  selectedDPath,
+  handleScanMoreAddresses,
+  onUnlock,
   handleUpdate
 }: DeterministicAccountListProps) {
+  const MAX_EMPTY_ADDRESSES = 5;
   const { isMobile } = useScreenSize();
 
-  const [selectedAccounts, setSelectedAccounts] = useState([] as ISelectedAccount[]);
-
-  const accountsToUse = uniqBy(prop('address'), finishedAccounts).filter(
-    ({ isFreshAddress, balance }) => (balance && !balance.isZero()) || isFreshAddress
-  );
+  const [tableAccounts, setTableAccounts] = useState({} as ITableAccounts);
 
+  const accountsToUse = uniqBy(prop('address'), finishedAccounts);
+  // setTableAccounts to be accountsToUse on update with isDefault set if it isn't already set and
   useEffect(() => {
-    const selected = uniqBy(
-      prop('address'),
-      accountsToUse.map(({ address, pathItem }) => ({
-        address,
-        derivationPath: pathItem.path
-      })) as ISelectedAccount[]
-    );
-    setSelectedAccounts(selected);
-  }, [accountsToUse.length]);
+    if (accountsToUse.length === 0 && Object.keys(tableAccounts).length !== 0) {
+      setTableAccounts({} as ITableAccounts);
+    } else {
+      const tableAccs = accountsToUse.reduce((acc, idx) => {
+        acc[idx.address] = tableAccounts[idx.address] || {
+          ...idx,
+          isDefaultConfig: true,
+          isSelected: (idx.balance && !idx.balance.isZero()) || false
+        };
+        return acc;
+      }, tableAccounts);
+      setTableAccounts(tableAccs);
+    }
+  }, [accountsToUse]);
 
+  const selectedAccounts = Object.values(tableAccounts).filter(({ isSelected }) => isSelected);
+  const emptySelectedAccounts = filterZeroBalanceAccounts(selectedAccounts);
   const handleSubmit = () => {
     onUnlock(selectedAccounts);
   };
 
-  const handleSelection = (account: DWAccountDisplay) => {
-    const newSelectedAccount: ISelectedAccount = {
-      address: account.address,
-      derivationPath: account.pathItem.path
-    };
-    const isPresent = selectedAccounts.find(({ address }) =>
-      isSameAddress(newSelectedAccount.address, address)
-    );
-    setSelectedAccounts(
-      isPresent
-        ? selectedAccounts.filter(
-            ({ address }) => !isSameAddress(newSelectedAccount.address, address)
-          )
-        : [...selectedAccounts, newSelectedAccount]
-    );
+  const handleSelection = (account: TableAccountDisplay) => {
+    if (account.isSelected) {
+      setTableAccounts({
+        ...tableAccounts,
+        [account.address]: {
+          ...account,
+          isDefaultConfig: false,
+          isSelected: !account.isSelected
+        }
+      });
+      return;
+    }
+    if (emptySelectedAccounts.length >= MAX_EMPTY_ADDRESSES) return;
```

Comment:
Will a newly found account be added automatically since this assumes that you are only selecting empty accounts manually? Is this also the case if my account is not found initially and I have to keep scanning?

Reply:
Yep, left a comment there explaining this.